thunder/core/proxies.py

```diff
@@ -1880,6 +1880,125 @@ def real(self):
         return method(self)
 
 
+class SubclassTensorProxy(TensorProxy):
+    _tensors: list[TensorProxy]
+    _non_tensors: list[Any]
+    _subclass_type: torch._C._TensorMeta
+
+    def __init__(self, *args, **kwargs):
+        from thunder.core.pytree import tree_flatten
+
+        kwarg_tensors = kwargs.pop("tensors", [])
+        kwarg_non_tensors = kwargs.pop("non_tensors", [])
+        subclass_type = kwargs.pop("subclass_type", None)
+
+        # If tensors (and non_tensors) are not empty, then it should be the path of `_make_wrapper_subclass`
+        # where `self` should already have gotten its name.
+        flat_args, spec = tree_flatten((args, kwargs))
+        tensors = list(filter(lambda t: isinstance(t, TensorProxy), flat_args))
+        non_tensors = list(filter(lambda t: not isinstance(t, TensorProxy), flat_args))
+        has_name_before_init = hasattr(self, "_name")
+
+        is_dunder_init_following_make_wrapper_subclass: bool = False
+        if tensors:
+            baseutils.check(
+                has_name_before_init
+                and not kwarg_tensors
+                and not kwarg_non_tensors
+                and self._subclass_type is not None,
+                lambda: (
+                    f"{flat_args=} indicates this instance is created by"
+                    "`torch.Tensor._make_wrapper_subclass`'s lookaside but `name` is not set"
+                ),
+            )
+            is_dunder_init_following_make_wrapper_subclass = True
+
+        if not is_dunder_init_following_make_wrapper_subclass:
+            super().__init__(*args, **kwargs)
+
+            self._tensors = kwarg_tensors
+            self._non_tensors = kwarg_non_tensors
+            self._subclass_type = subclass_type
+        else:
+            # TODO(crcrpar): Think about materializing `self` so that we can
+            # call `__tensor_init__` to know each attribute names.
+            from thunder.core import prims
+
+            self._tensors = tensors
+            self._non_tensors = non_tensors
+            bsym = prims.tensor_subclass_ctor.bind(
```

Comment:
This is interesting. Why does this class either create a new proxy or change the actual trace?

Reply:
this path is happening while tracing and the call of `Proxy.__init__` is not recorded by default in a trace so it's necessary to deliberately register a boundsymbol to the currently active scope

Reply:
OK; but why does proxy creation get recorded into the trace?

Reply:
if a function creates a subclass inside it then shouldn't a trace of it have a BoundSymbol to represent it?

Reply:
I think it's OK to put tensor subclass creation into a trace (although I'm curious if we can dce the creation if the subclass is just flattened and used once later), but I'm not sure why the operator that creates the actual tensor subclass is also a constructor for the proxy. The existing tensor proxies don't entangle their tensor factory methods with the creation of the proxy, for example

Reply:
The short answer would be that it's discouraged to use `torch.Tensor(...)` but we do `MySubclass(...)` and `MySubclass.__new__` is visible to Thunder.

`MySubclass(...)` (in this PR, it's a subclass that calls `_make_wrapper_subclass` in its `__new__`) calls `__new__`. thunder's tracing sees this dunder new. `_make_wrapper_subclass` used in this method requires a new lookaside or a new torchsymbol (Both would be supposed to return a proxified instance of `MySubclass`). Dunder init of `MySubclass` is called on the return value of `MySubclass.__new__`.
But here, as we override the behavior of `MySubclass.__new__`, the init of the proxified `MySubclass` instance is called.

Reply:
In the end, a trace might end up free from bsyms representing `MySubclass(...)`, currently this is necessary to just get an initial trace.

Reply:
I still don't really get this. I understand that Thunder sees the call to `__new__`, I'm just confused why that call to `__new__` doesn't produce a bound symbol like other lookasides, and then the meta function for that bound symbol produces a SubclassTensorProxy?

I wouldn't say that we "discourage" the use of `torch.Tensor` (although I don't think Thunder supports it, so I guess that's discouraging) any more than PyTorch does. Thunder does support `torch.tensor`, however, and that does directly call the TensorProxy constructor

Reply:
As for `torch.Tensor(...)`, I should've put the link to this page: https://pytorch.org/docs/stable/tensors.html#torch.Tensor.__init__.
> This constructor is deprecated, we recommend using [torch.tensor()](https://pytorch.org/docs/stable/generated/torch.tensor.html#torch.tensor) instead. What this constructor does depends on the type of data.

> why that call to __new__ doesn't produce a bound symbol like other lookasides, and then the meta function for that bound symbol produces a SubclassTensorProxy?

Would this by chance mean that a lookaside for `__new__` should work better than a lookaside for `torch.Tensor._make_wrapper_subclass`?
From my perspective, either way, as we see `__new__`, then `__init__` is to be called so I guess it'd be necessary to do some meticulous handling of `__init__` of tensor subclasses because either way we won't dodge the call of the dunder new.

Reply:
OK; I think I'm understanding better what the issue is. Let me try to summarize it. 

When someone creates a tensor subclass, it looks like this:

```
x = ComplexTensor(torch.tensor([[1]]), torch.tensor([[2]]))
```

(taken from https://github.com/albanD/subclass_zoo/blob/main/complex_tensor.py)

And that makes sense. The class's constructor is called, because of course it is.

Internally, Python looks at the class constructor and then constructs an object of the requested type (this process is a little complicated), then initializes it (optionally) using dunder init.

How we understand the object created by the call to `ComplexTensor` is determined by the subclass's implementation of dunder torch_dispatch, dunder tensor flatten, and dunder tensor unflatten. Does the actual call to _make_wrapper_subclass tell us anything about the subclass, or do we get that information by querying the object returned from the constructor?

I guess my question, then, is if we can also just call `ComplexTensor` like a practitioner would? Then the meta for `ComplexTensor` could return a `SubclassTensorProxy` or even a `ComplexTensorSubclassProxy`.

I think that would provide a clear separation between the proxy and the actual call.

Reply:
> Does the actual call to `_make_wrapper_subclass` tell us anything about the subclass

`torch.Tensor._make_wrapper_subclass` takes `cls`. The lookaside could let us tell things other than actual tensors and other values to be registered to the output of the method.

> I guess my question, then, is if we can also just call `ComplexTensor` like a practitioner would? 

I don't get where this call of `ComplexTensor` would be. Could you elaborate on it?

> Then the meta for `ComplexTensor` could return a `SubclassTensorProxy` or even a `ComplexTensorSubclassProxy`.

How and when would we register `ComplexTensorSubclassProxy`? Would this indicate that Thunder is going to detect a subclass and define a proxy for it dynamically?

Reply:
Yes! What if we detected that a subclass constructor was called and dynamically created a symbol for the constructor and the corresponding proxy?

The call to `ComplexTensor` could be in the resulting Thunder program, like other bound symbols. Or maybe we want to represent `ComplexTensor` construction as an unflatten call, like `ComplexTensor.__tensor_unflatten__`? Both seem reasonable to me

Reply:
If I do that dynamically, I want that subclass proxy to correctly capture the result type of ops with that subclass. I however don't see a clear path to infer the result type of ops with subclasses, other than we do what AOT Autograd does with torch.fx tracing. Thus I don't think dynamic registration is that different from what I have in this PR.

Representing subclass construction as unflatten call would then call `__init__` of that subclass, so I'm not following the point.

Reply:
The goal is to divorce the proxy from the actual runtime object to simplify the code and its concepts.

How else would we infer the decomposition of operations called on a tensor subclass, and if the result is also a member of the subclass? 

I understand that __init__ gets called on the object at runtime, of course. The goal is not to someone circumvent the proper construction of the object at runtime (unless it can be elided for performance).

Reply:
Maybe I'm not understanding this well, but I'm a bit worried that we don't have well-defined semantics here and that the traces are not representing what's up.

We should absolutely know when we want to construct and use the subclass object (if we return it or something wants the subclass object) and when we don't (all other cases) and should also represent what will be the compute in the trace.

We also want to minimize admin overhead during the compute, which we are not doing a great job today (having looked at wall clock vs. GPU self time for Llama 1b today), so adding the overhead of dealing with subclasses at compute time should likely be minimized.

Reply:
excuse me for my idling on this. I guess I understand your points a bit better. The separation does sound nice. For it, I think one way is to do the interpretation of `__torch_dispatch__` while acquiring the initial trace, not what #1394 is doing, i.e., get the initial trace with opaque tensor subclass proxies before unrolling their `__torch_dispatch__`. With this, some traces like ones we'd get from torchao float8 programs would be completely free from tensor subclass proxies. If any of arguments or return values of a function is a tensor subclass, we might end up having one or more proxy classes for tensor subclasses observed in it and we might need to implement a dynamic registration of them while interpretation, which I almost have no idea at the moment.

Reply:
Happy to talk through this more over VC if it would be helpful, too!

Reply:
Tensor subclass proxy objects would have `type(tensor_subclass_object)` as their attribute but not actual instances.

I'd be rather hesitant to complicate the current source code of interpreter by introducing mechanism which unrolls tensor subclasses, defines a proxy class for the observed tensor subclass, and registers that proxy class to the namespace.
Currently, the implementation itself isn't the best for sure, but the unrolling of tensor subclasses happens after the initial computation is acquired (This initial trace would have some wrong expressions, e.g., when a tensor subclass' `__torch_dispatch__` returns one or multiple objects of that tensor subclass), thus I think the change of this PR (more precisely, #1394 and #1415) would have less interactions.